woocommerce/woocommerce
# [COT] Implement metadata CRUD and `update()` method in COT datastore

### Description

This PR takes a first stab at implementing the `update()` function for the COT datastore, which would allow saving of existing orders via `$order->save()`.

Given metadata is currently handled by `WC_Data_Store_WP` via `Abstract_WC_Order_Data_Store_CPT`, which are unaware of the table structure of the new `wc_orders_meta` table and instead expect a traditional WP metadata table, I've added overrides to the datastore which relay the calls to a new helper class `OrdersTableDataStoreMeta`.

Similarly, given `update()` can no longer use `wp_update_post()` or regular metadata, the code relies on `OrdersTableDataStoreHelper` to provide some functionality that keeps the implementation a bit more clean, by generating the correct data and format that `$wpdb` needs for updating the different COT tables.

Closes #32667.

### Testing

1. Enable the COT feature.
2. Migrate some orders over from the posts table to COT tables.
3. Make sure `OrdersTableDataStore` datastore is the datastore in use. This can be done through the UI or "forcefully" with code along these lines:
   ```php
   add_filter(
	   'woocommerce_order_data_store',
	   function ( $default_data_store ) {
		   return wc_get_container()->get( 'Automattic\WooCommerce\Internal\DataStores\Orders\OrdersTableDataStore' );
	   },
	   999,
	   1
   );
   ```
4. Via custom code, load an order and try to read/modify some properties, add metadata, etc and then save. For example:

   ```php
   $order = wc_get_order( 123 ); // Use a valid order ID.
   $order->set_billing_first_name( 'Nicholai' );
   $order->set_billing_email( 'nicholai.hel@shibumi.test' );
   $order->set_total( 12.0 );
   $order->set_transaction_id( 'XX123456' );
   $order->set_prices_include_tax( true );
   $order->add_meta_data( 'my_own_meta', 123 );
   $order->save();
   
   echo $order->get_meta( '_shipping_city' );
   ```
5. Confirm that tables `wc_orders`, `wc_order_operational_data`, `wc_order_addresses` and `wc_orders_meta` have been updated accordingly.

### Considerations/discussion

- There are a bunch of TODO's with things still to be done.
- I'm using `OrdersTableDataStoreMeta` as a helper that knows about the `wc_orders_meta` structure and to which `OrdersTableDataStoreMeta` passes all metadata-related requests, but that only to speed up progress. Ideally we would turn this into a more abstract class that could be inherited by `OrdersTableDataStore` and hypothetical future datastores that might custom meta tables, similarly to how `WC_Data_Store_WP` can be used by any datastore using WP's metadata functions.
- `OrdersTableDataStoreHelper` duplicates code that is found in various places dealing with the database (such as some migration classes). It'd probably be a good idea to "unify" column mappings, data type conversions, etc. for COT in a single place that all other classes could use.
- Updates in the different tables are performed sequentially (for easier debugging), but we could probably do it in just one pass.
- The datastore in general (and the `update()` in particular) don't yet handle "internal datastore" properties such as `_download_permissions_granted`, `_new_order_email_sent`, etc. correctly. I was planning on borrowing some code from #32857 for that. This is the main reason I'm marking this PR as a _draft_ for now.
